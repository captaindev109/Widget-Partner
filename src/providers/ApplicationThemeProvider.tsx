import { ThemeProvider, DefaultTheme } from 'styled-components';
import GlobalStyles from '@/styles/Global';
import React from 'react';
import { ChildrenReturnType } from '@/types/types';

interface ApplicationThemeProviderProps {
  children?: ChildrenReturnType;
  theme: DefaultTheme;
}

const ApplicationThemeProvider = ({ children, theme }: ApplicationThemeProviderProps) => {
  return (
    <ThemeProvider theme={theme}>
      <GlobalStyles />
      {children}
    </ThemeProvider>
  );
};

export default ApplicationThemeProvider;
