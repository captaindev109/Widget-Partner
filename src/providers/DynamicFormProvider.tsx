import React from 'react';
import { ChildrenReturnType } from '@/types/types';
import { FormProvider, useForm } from 'react-hook-form';

interface DynamicFormProviderProps {
  children?: ChildrenReturnType;
}

const useDynamicRootForm = () => {
  const methods = useForm({ mode: 'onChange' });

  return { methods };
};

export const DynamicFormProvider = ({ children }: DynamicFormProviderProps) => {
  const { methods } = useDynamicRootForm();

  return <FormProvider {...methods}>{children}</FormProvider>;
};
