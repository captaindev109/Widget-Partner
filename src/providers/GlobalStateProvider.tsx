import { mainMachine } from '@/machines/courseMachine';
import headerMachine from '@/machines/headerMachine';
import footerMachine from '@/machines/footerMachine';
import React, { createContext } from 'react';
import { useInterpret } from '@xstate/react';
import { ActorRefFrom } from 'xstate';
import { storeMachine } from '@/machines/storeMachine';

interface GlobalStateContextProps {
  mainService: ActorRefFrom<typeof mainMachine>;
  headerService: ActorRefFrom<typeof headerMachine>;
  footerService: ActorRefFrom<typeof footerMachine>;
  storeService: ActorRefFrom<typeof storeMachine>;
}

export const GlobalStateContext = createContext({} as GlobalStateContextProps);

interface GlobalStateProviderProps {
  children?: React.ReactNode;
}

const useDevTool = process.env.NODE_ENV === 'development';

export const GlobalStateProvider = ({ children }: GlobalStateProviderProps) => {
  const services = {
    mainService: useInterpret(mainMachine, {
      devTools: useDevTool,
    }),
    headerService: useInterpret(headerMachine, {
      devTools: useDevTool,
    }),
    footerService: useInterpret(footerMachine, {
      devTools: useDevTool,
    }),
    storeService: useInterpret(storeMachine, {
      devTools: useDevTool,
    }),
  };

  return <GlobalStateContext.Provider value={services}>{children}</GlobalStateContext.Provider>;
};
