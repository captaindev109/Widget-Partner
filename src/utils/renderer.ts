import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
