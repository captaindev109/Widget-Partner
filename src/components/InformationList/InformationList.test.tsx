/**
 *
 *  ********** Unit & Snapshot Testing *********
 *
 * Unit test is specifiic for each functionliaty of the component
 * Snapshot test is to track Over All changes in the code since the first snap
 **/

import { render, RenderResult } from '@testing-library/react';
import queries from '@testing-library/dom/types/queries';
import renderer from 'react-test-renderer';
import { InformationItem, InformationList } from '.';
import { GlobalStateProvider } from '@/providers/GlobalStateProvider';
import ApplicationThemeProvider from '@/providers/ApplicationThemeProvider';
import defaultTheme from '@/theme/theme';

const title = 'Title';
const price = '1 000 000 kr';
const description = 'This is a description';

const MockComponent: React.FC = () => {
  return (
    <GlobalStateProvider>
      <ApplicationThemeProvider theme={defaultTheme}>
        <InformationItem title={title} price={price} description={description} />
      </ApplicationThemeProvider>
    </GlobalStateProvider>
  );
};

// describe function is a better way to organize our test cases and describe which component we are testing
describe('<InformationItem /> Test Cases', () => {
  let renderedComponent: RenderResult<typeof queries, HTMLElement>;
  beforeEach(() => {
    renderedComponent = render(<MockComponent />);
  });

  // this test is a must for each component, any rendering issues with the component can be found
  test('it renders without crashing, all is well', () => {
    // render basically prints and exposes the component in nodejs environemt, it returns an object...
    render(<MockComponent />);
  });

  test('check to see any specific string, element in dom, for example', () => {
    const { getByText } = renderedComponent;

    // YOU DO NOT NEED TO USE ALL OF THESE CASES FOR EVERY COMPONENT, JUST ONE OR MORE IS ENOUGH

    // using string value
    expect(getByText(title)).toBeInTheDocument();
    expect(getByText(price)).toBeInTheDocument();
    expect(getByText(description)).toBeInTheDocument();

    //using selectors
  });

  it('Snapshot test should pass', () => {
    const snap = renderer.create(<MockComponent />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});

const MockListComponent: React.FC = () => {
  return (
    <GlobalStateProvider>
      <ApplicationThemeProvider theme={defaultTheme}>
        <InformationList>
          <InformationItem title={title} price={price} description={description} />
          <InformationItem title={title} price={price} description={description} />
          <InformationItem title={title} price={price} description={description} />
        </InformationList>
      </ApplicationThemeProvider>
    </GlobalStateProvider>
  );
};

// describe function is a better way to organize our test cases and describe which component we are testing
describe('<InformationList /> Test Cases', () => {
  let renderedComponent: RenderResult<typeof queries, HTMLElement>;
  beforeEach(() => {
    renderedComponent = render(<MockListComponent />);
  });

  // this test is a must for each component, any rendering issues with the component can be found
  test('it renders without crashing, all is well', () => {
    // render basically prints and exposes the component in nodejs environemt, it returns an object...
    render(<MockListComponent />);
  });

  test('check to see any specific string, element in dom, for example', () => {
    const { container } = renderedComponent;

    // YOU DO NOT NEED TO USE ALL OF THESE CASES FOR EVERY COMPONENT, JUST ONE OR MORE IS ENOUGH

    // using string value
    expect(container.firstChild).toBeInTheDocument();

    //using selectors
  });

  it('Snapshot test should pass', () => {
    const snap = renderer.create(<MockListComponent />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
