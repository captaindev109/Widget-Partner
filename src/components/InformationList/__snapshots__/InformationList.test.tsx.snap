// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InformationItem /> Test Cases Snapshot test should pass 1`] = `
<div
  className="Card-sc-fb64gk-0 InformationList__StyledInformationItem-sc-1i41yqw-0 iJgSZC iFyMa-D"
  data-testid="InformationItem"
>
  <p
    className="Text__StyledText-sc-1jfsuh1-0 lmAVHO"
    color="gray700"
  >
    Title
  </p>
  <p
    className="Text__StyledText-sc-1jfsuh1-0 iEvncp"
    color="gray900"
  >
    1 000 000 kr
  </p>
  <div
    className="Card-sc-fb64gk-0 InformationList__StyledCard-sc-1i41yqw-1 leCJCh fhYwwt"
  >
    <div
      className="Card-sc-fb64gk-0 Tooltip__TooltipCard-sc-xggov6-2 leCJCh hfUwNk"
    >
      <div
        className="Card-sc-fb64gk-0 Tooltip__TooltipIcon-sc-xggov6-1 leCJCh cyLRGr"
      >
        <svg
          fill="none"
          height="20px"
          viewBox="0 0 20 20"
          width="20px"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M9.25 15.5V8.5H10.75V15.5H9.25Z"
            fill="#121212"
          />
          <path
            d="M10 6.875C10.6904 6.875 11.25 6.31536 11.25 5.625C11.25 4.93464 10.6904 4.375 10 4.375C9.30964 4.375 8.75 4.93464 8.75 5.625C8.75 6.31536 9.30964 6.875 10 6.875Z"
            fill="#121212"
          />
          <path
            clipRule="evenodd"
            d="M10 20C15.5228 20 20 15.5228 20 10C20 4.47715 15.5228 0 10 0C4.47715 0 0 4.47715 0 10C0 15.5228 4.47715 20 10 20ZM10 18.5C14.6944 18.5 18.5 14.6944 18.5 10C18.5 5.30558 14.6944 1.5 10 1.5C5.30558 1.5 1.5 5.30558 1.5 10C1.5 14.6944 5.30558 18.5 10 18.5Z"
            fill="#121212"
            fillRule="evenodd"
          />
        </svg>
      </div>
      <div
        className="Card-sc-fb64gk-0 Tooltip__TooltipText-sc-xggov6-0 leCJCh fDlHgV"
      >
        <p
          className="Text__StyledText-sc-1jfsuh1-0 cJpxlc"
          color="gray100"
        >
          This is a description
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`<InformationList /> Test Cases Snapshot test should pass 1`] = `
<div
  className="Card-sc-fb64gk-0 InformationList__StyledInformationList-sc-1i41yqw-2 leCJCh kYtJuX"
  data-testid="InformationList"
>
  <div
    className="Card-sc-fb64gk-0 InformationList__StyledInformationItem-sc-1i41yqw-0 iJgSZC iFyMa-D"
    data-testid="InformationItem"
  >
    <p
      className="Text__StyledText-sc-1jfsuh1-0 lmAVHO"
      color="gray700"
    >
      Title
    </p>
    <p
      className="Text__StyledText-sc-1jfsuh1-0 iEvncp"
      color="gray900"
    >
      1 000 000 kr
    </p>
    <div
      className="Card-sc-fb64gk-0 InformationList__StyledCard-sc-1i41yqw-1 leCJCh fhYwwt"
    >
      <div
        className="Card-sc-fb64gk-0 Tooltip__TooltipCard-sc-xggov6-2 leCJCh hfUwNk"
      >
        <div
          className="Card-sc-fb64gk-0 Tooltip__TooltipIcon-sc-xggov6-1 leCJCh cyLRGr"
        >
          <svg
            fill="none"
            height="20px"
            viewBox="0 0 20 20"
            width="20px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.25 15.5V8.5H10.75V15.5H9.25Z"
              fill="#121212"
            />
            <path
              d="M10 6.875C10.6904 6.875 11.25 6.31536 11.25 5.625C11.25 4.93464 10.6904 4.375 10 4.375C9.30964 4.375 8.75 4.93464 8.75 5.625C8.75 6.31536 9.30964 6.875 10 6.875Z"
              fill="#121212"
            />
            <path
              clipRule="evenodd"
              d="M10 20C15.5228 20 20 15.5228 20 10C20 4.47715 15.5228 0 10 0C4.47715 0 0 4.47715 0 10C0 15.5228 4.47715 20 10 20ZM10 18.5C14.6944 18.5 18.5 14.6944 18.5 10C18.5 5.30558 14.6944 1.5 10 1.5C5.30558 1.5 1.5 5.30558 1.5 10C1.5 14.6944 5.30558 18.5 10 18.5Z"
              fill="#121212"
              fillRule="evenodd"
            />
          </svg>
        </div>
        <div
          className="Card-sc-fb64gk-0 Tooltip__TooltipText-sc-xggov6-0 leCJCh fDlHgV"
        >
          <p
            className="Text__StyledText-sc-1jfsuh1-0 cJpxlc"
            color="gray100"
          >
            This is a description
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="Card-sc-fb64gk-0 InformationList__StyledInformationItem-sc-1i41yqw-0 iJgSZC iFyMa-D"
    data-testid="InformationItem"
  >
    <p
      className="Text__StyledText-sc-1jfsuh1-0 lmAVHO"
      color="gray700"
    >
      Title
    </p>
    <p
      className="Text__StyledText-sc-1jfsuh1-0 iEvncp"
      color="gray900"
    >
      1 000 000 kr
    </p>
    <div
      className="Card-sc-fb64gk-0 InformationList__StyledCard-sc-1i41yqw-1 leCJCh fhYwwt"
    >
      <div
        className="Card-sc-fb64gk-0 Tooltip__TooltipCard-sc-xggov6-2 leCJCh hfUwNk"
      >
        <div
          className="Card-sc-fb64gk-0 Tooltip__TooltipIcon-sc-xggov6-1 leCJCh cyLRGr"
        >
          <svg
            fill="none"
            height="20px"
            viewBox="0 0 20 20"
            width="20px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.25 15.5V8.5H10.75V15.5H9.25Z"
              fill="#121212"
            />
            <path
              d="M10 6.875C10.6904 6.875 11.25 6.31536 11.25 5.625C11.25 4.93464 10.6904 4.375 10 4.375C9.30964 4.375 8.75 4.93464 8.75 5.625C8.75 6.31536 9.30964 6.875 10 6.875Z"
              fill="#121212"
            />
            <path
              clipRule="evenodd"
              d="M10 20C15.5228 20 20 15.5228 20 10C20 4.47715 15.5228 0 10 0C4.47715 0 0 4.47715 0 10C0 15.5228 4.47715 20 10 20ZM10 18.5C14.6944 18.5 18.5 14.6944 18.5 10C18.5 5.30558 14.6944 1.5 10 1.5C5.30558 1.5 1.5 5.30558 1.5 10C1.5 14.6944 5.30558 18.5 10 18.5Z"
              fill="#121212"
              fillRule="evenodd"
            />
          </svg>
        </div>
        <div
          className="Card-sc-fb64gk-0 Tooltip__TooltipText-sc-xggov6-0 leCJCh fDlHgV"
        >
          <p
            className="Text__StyledText-sc-1jfsuh1-0 cJpxlc"
            color="gray100"
          >
            This is a description
          </p>
        </div>
      </div>
    </div>
  </div>
  <div
    className="Card-sc-fb64gk-0 InformationList__StyledInformationItem-sc-1i41yqw-0 iJgSZC iFyMa-D"
    data-testid="InformationItem"
  >
    <p
      className="Text__StyledText-sc-1jfsuh1-0 lmAVHO"
      color="gray700"
    >
      Title
    </p>
    <p
      className="Text__StyledText-sc-1jfsuh1-0 iEvncp"
      color="gray900"
    >
      1 000 000 kr
    </p>
    <div
      className="Card-sc-fb64gk-0 InformationList__StyledCard-sc-1i41yqw-1 leCJCh fhYwwt"
    >
      <div
        className="Card-sc-fb64gk-0 Tooltip__TooltipCard-sc-xggov6-2 leCJCh hfUwNk"
      >
        <div
          className="Card-sc-fb64gk-0 Tooltip__TooltipIcon-sc-xggov6-1 leCJCh cyLRGr"
        >
          <svg
            fill="none"
            height="20px"
            viewBox="0 0 20 20"
            width="20px"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M9.25 15.5V8.5H10.75V15.5H9.25Z"
              fill="#121212"
            />
            <path
              d="M10 6.875C10.6904 6.875 11.25 6.31536 11.25 5.625C11.25 4.93464 10.6904 4.375 10 4.375C9.30964 4.375 8.75 4.93464 8.75 5.625C8.75 6.31536 9.30964 6.875 10 6.875Z"
              fill="#121212"
            />
            <path
              clipRule="evenodd"
              d="M10 20C15.5228 20 20 15.5228 20 10C20 4.47715 15.5228 0 10 0C4.47715 0 0 4.47715 0 10C0 15.5228 4.47715 20 10 20ZM10 18.5C14.6944 18.5 18.5 14.6944 18.5 10C18.5 5.30558 14.6944 1.5 10 1.5C5.30558 1.5 1.5 5.30558 1.5 10C1.5 14.6944 5.30558 18.5 10 18.5Z"
              fill="#121212"
              fillRule="evenodd"
            />
          </svg>
        </div>
        <div
          className="Card-sc-fb64gk-0 Tooltip__TooltipText-sc-xggov6-0 leCJCh fDlHgV"
        >
          <p
            className="Text__StyledText-sc-1jfsuh1-0 cJpxlc"
            color="gray100"
          >
            This is a description
          </p>
        </div>
      </div>
    </div>
  </div>
</div>
`;
